src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,185 @@
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass.ToString() == RequiresPreviewFeaturesAttribute);
+                if (!(attribute is null))
+                {
+                    // This assembly has enabled preview attributes.
+                    return;
+                }
+
+                var enabledPreviewFeatures = context.Options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.EnablePreviewFeatures, context.Compilation, context.CancellationToken);
+                if (string.Equals(enabledPreviewFeatures, "true", System.StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols = new ConcurrentDictionary<ISymbol, IOperation?>();
+                context.RegisterOperationAction(context => BuildSymbolInformation(context, requiresPreviewFeaturesSymbols),
+                    OperationKind.Invocation,
+                    OperationKind.ObjectCreation,
+                    OperationKind.PropertyReference,
+                    OperationKind.FieldReference,
+                    OperationKind.DelegateCreation,
+                    OperationKind.EventReference
+                    );
+            });
+        }
+
+        private void BuildSymbolInformation(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            if (OperationUsesPreviewFeatures(context.Operation, requiresPreviewFeaturesSymbols))
+            {
+                context.ReportDiagnostic(context.Operation.CreateDiagnostic(Rule));
+            }
+        }
+
+        private bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            ISymbol? symbol = GetOperationSymbol(operation);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            return OperationUsesPreviewFeatures(operation, symbol, requiresPreviewFeaturesSymbols, true);
+        }
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
+                IFieldReferenceOperation fOperation => fOperation.Field,
+                IDelegateCreationOperation dOperation => dOperation.Type,
+                IEventReferenceOperation eOperation => eOperation.Member,
+                _ => null,
+            };
+
+        private bool TryGetOrCachePreviewFeaturesAttributesOnSymbol(IOperation operation, ISymbol symbol, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols, bool checkParents)
```

Comment:
Looks like `checkParents` is always true?

Reply:
For now. I'm keeping it around because I expect we will decide to stop checking all the way up to the root when we start expanding the test cases.